### DISCUSSION TITLE
Support for batch operations

### DISCUSSION THREAD

**Discussion by fn-yves** [OpenAssetIO/OpenAssetIO#37] [Ideas and change proposals]
# Support for batch operations

The question of how best to support batch operations within OAIO came up recently. We took a first pass at this some years ago, and wanted to note our thinking on this. How does this fit with the way people would want to implement batching in a manager plugin - and are there any better ideas?

## Strategy 1: Plural API Methods (current approach)
Our current approach to enabling batch operations is to implement a plural equivalent of a given method. For example, the method `resolveEntityReference` might have a plural equivalent `resolveEntityReferences`.

```python
def resolveEntityReference(self, reference, context):
def resolveEntityReferences(self, references, context):
```

**For managers**, an advantage of this approach is that it is single-instruction, multiple-data (SIMD), which simplifies implementation because each batch operation only has a single context. Further, a single context for a given operation also simplifies caching, making optimization easier.

**For hosts**, this is still a useful pattern, but does require structural changes to control flow to 
make best use of it.

However, this approach has the disadvantage that it precludes performing different types of operations within the same batch. If this is a major drawback, then we need to consider other strategies for performing batch operations.

## Strategy 2: Command Architecture (considered but rejected)
Instead of having separate API methods for each operation, we investigated using a command architecture where each command has an identifier (similar to protobuf) and some arguments. Batches here become multiple-instruction, multiple-data (MIMD). 

**For managers**, this potentially allows operations of different types to be batched and satisfied with fewer back-end queries. For example a mix of `resolve` and `getMetadata`. 

This provides maximal flexibility for the manager to optimise the batch however possible, but at the expense of a somewhat less conventional API.

It also requires significant manual leg-work on the manager side to determine the actual batchable groups of commands vs those that were supplied (due to the context or other params, etc..). There is a potential complexity explosion if the context is considered. It seems like any batched command architecture should have the constraint that only a single context can be used for the whole batch.

**For hosts**, this approach has similar disadvantages in that it also requires the same control flow restructuring as in many situations, the result of a previous method is often required in order to make the next API call.

The main perceived downside of this approach though, is that for either side of the API, this architecture would introduce significant complexity in the implementation. Especially in statically typed languages, and more generally the broader usability and familiarity of the API would suffer.

Because of this, we opted for the simpler/more familiar route of adding plural versions of a subset of the API methods that are commonly queried in performance sensitive scenarios.

**Comment by pshinners-blz** [comment]
This matches the way our asset system is implemented. On our servers, ever call is implemented as a "plural call". Every call takes a list of nouns to operate, and they all run with the exact same arguments. On the client side, each we generate a "singular" and "plural" variant of the same function that are directed to the same implementation on the server.

Last I looked, the base classes for OAIO managers are implemented the wrong way; the default batch calls call through to the singular calls. If this was switched, anyone implementing a manager would only need to fill out the batch or the singular call depending on how their asset system worked. With the current design I need to implement both singular and plural calls to work with our batching asset system. (I'm pretty sure this is correct, if so I strongly recommend changing the default implementations)

**Reply by foundrytom** [reply]
Thanks Pete, that makes a lot of sense, and is an easy switch. I've made #43 to cover this - it seems like a no-brainer.

**Comment by pshinners-blz** [comment]
I forgot error handling. In the case of batch lookups, there must be some return value that can be returned to indicate "the value for this index is not value"

**Reply by foundrytom** [reply]
So is the expected behaviour of a batch, that you shouldn't fail the batch ever?

Eg, right now, in the singular calls, various exceptions are used to indicate error state and allow easy propagation of messages, etc.. In the batch version, should this be effectively returned as part of the result instead, so that `resolveReferences` should never throw? [edit: for per-value errors]

**Comment by pshinners-blz** [comment]
I could imagine only the call to validate assetIds are the only functions that can fail with interesting results. The other functions may only need to respond with, "that doesn't actually exist"? That may be too optimistic, but simplifies the design if it works. There is still the problem of different functions returning different types (strings, hashmaps, booleans) and each manager function knowing how to populate values that weren't found.

I wouldn't want the entire manager method to have to fail if one of N lookups didn't have the expected results.

**Comment by feltech** [comment]
Hey all, I have updated #43 with a proposal to replace all `ManagerInterface` methods with batch variants, such that singular variants only exist in a convenience wrapper, and hence manager plugins only need implement the batch variant.  Would be great to get some feedback in that issue. Thanks.

**Comment by foundrytom** [comment]
https://github.com/TheFoundryVisionmongers/OpenAssetIO/pull/98 has now been merged.
